daft/execution/dynamic_schedule.py

```diff
@@ -0,0 +1,499 @@
+from __future__ import annotations
+
+import math
+from abc import abstractmethod
+from collections.abc import Iterator
+from dataclasses import dataclass, field
+from typing import Callable, Optional
+
+from daft.execution.dynamic_construction import (
+    Construction,
+    make_fanout_range_instruction,
+    make_join_instruction,
+    make_local_limit_instruction,
+    make_map_to_samples_instruction,
+    make_merge_and_sort_instruction,
+    make_merge_instruction,
+    make_read_file_instruction,
+    make_reduce_to_quantiles_instruction,
+    make_write_instruction,
+)
+from daft.logical import logical_plan
+from daft.runners.partitioning import PartitionSet, vPartition
+from daft.runners.pyrunner import LocalPartitionSet
+
+
+class DynamicSchedule(Iterator[Optional[Construction]]):
+    """Recursively dynamically generate a sequence of execution steps to compute some target.
+
+    - The execution steps (Constructions) are exposed via the iterator interface.
+    - If the Construction is marked for materialization, it should be materialized and reported as completed.
+    - If None is returned, it means that the next execution step is waiting for some previous Construction to materialize.
+    """
+
+    @dataclass
+    class Materializations:
+        dependencies: list[vPartition | None] = field(default_factory=list)
+
+    def __init__(self) -> None:
+        # Materialized partitions that this dynamic schedule needs to do its job.
+        # `None` denotes a partition whose materialization is in-progress.
+        self._materializations: DynamicSchedule.Materializations = self.Materializations()
+        self._completed = False
+
+    def __next__(self) -> Construction | None:
+        if self._completed:
+            raise StopIteration
+
+        try:
+            return self._next_impl()
+
+        except StopIteration:
+            self._completed = True
+            # Drop references to dependencies.
+            self._materializations = self.Materializations()
+            raise
+
+    @abstractmethod
+    def _next_impl(self) -> Construction | None:
+        """
+        Raises StopIteration if there are no further instructions to give for this schedule.
+        Returns None if there is nothing we can do for now (i.e. must wait for other partitions to finish).
+        """
+        raise NotImplementedError()
+
+
+class SchedulePartitionRead(DynamicSchedule):
+    def __init__(self, partitions: list[vPartition]) -> None:
+        super().__init__()
+        self._materializations.dependencies += partitions
+        self._num_dispatched = 0
+
+    def _next_impl(self) -> Construction | None:
+
```

Comment:
I would consider making these generator functions as is so we don't have to track things like `self._num_dispatched ` or raise stop iteration
something like 
```
def _next_impl(self):
    yield from  (Construction( [partition]) for partition in _materializations.dependencies)
```